### PR TITLE
planner: remove some risky cache operations in the plan builder

### LINKED ISSUE #22850: runtime error: index out of range [2] with length 1
### 1. Minimal reproduce step (Required)
```sql
drop table if exists t1;
CREATE TABLE t1 (a int(11));
SELECT @v:=(SELECT 1 FROM t1 t2 LEFT JOIN t1 ON t1.a GROUP BY t1.a) FROM t1;
```

### 2. What did you expect to see? (Required)
No panic.

### 3. What did you see instead (Required)
Panic:
```log
[2021/02/20 18:31:16.184 +08:00] [ERROR] [conn.go:734] ["connection running loop panic"] [conn=5] [lastSQL="SELECT IF( @v:=LEAST((SELECT 1 FROM t1  t2 LEFT JOIN t1 ON (@v) GROUP BY t1.a), a), count(*), 1) FROM t1  GROUP BY a LIMIT 1"] [err="runtime error: index out of range [2] with length 1"] [stack="goroutine 982 [running]:\ngithub.com/pingcap/tidb/server.(*clientConn).Run.func1(0x3f8a0a0, 0xc001187950, 0xc000a2af00)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/server/conn.go:732 +0xee\npanic(0x3730fe0, 0xc00146e000)\n\t/usr/local/go/src/runtime/panic.go:679 +0x1b2\ngithub.com/pingcap/tidb/planner/core.(*expressionRewriter).toColumn(0xc0013c5380, 0xc0013af000)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/expression_rewriter.go:1725 +0xfba\ngithub.com/pingcap/tidb/planner/core.(*expressionRewriter).Leave(0xc0013c5380, 0x3fa6ac0, 0xc0013af000, 0x3fa6ac0, 0xc0013af000, 0x203000)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/expression_rewriter.go:1043 +0x1b42\ngithub.com/pingcap/parser/ast.(*ColumnName).Accept(0xc0013af000, 0x3f52260, 0xc0013c5380, 0x3fa6b20, 0xc0013c74d0, 0xc0009f4800)\n\t/home/jenkins/agent/workspace/tidb_master/go/pkg/mod/github.com/pingcap/parser@v0.0.0-20210203141130-32ef3e0e2e13/ast/expressions.go:527 +0x8a\ngithub.com/pingcap/parser/ast.(*ColumnNameExpr).Accept(0xc0013c74d0, 0x3f52260, 0xc0013c5380, 0xc0009f48c0, 0xc001452f50, 0x0)\n\t/home/jenkins/agent/workspace/tidb_master/go/pkg/mod/github.com/pingcap/parser@v0.0.0-20210203141130-32ef3e0e2e13/ast/expressions.go:589 +0x96\ngithub.com/pingcap/tidb/planner/core.(*PlanBuilder).rewriteExprNode(0xc001356a80, 0xc0013c5380, 0x3fc1e80, 0xc0013c74d0, 0xc0009f4801, 0x0, 0x0, 0x0, 0x0, 0x0, ...)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/expression_rewriter.go:195 +0x8e\ngithub.com/pingcap/tidb/planner/core.(*PlanBuilder).rewriteWithPreprocess(0xc001356a80, 0x3f8a0a0, 0xc001400840, 0x3fc1e80, 0xc0013c74d0, 0x3fdd980, 0xc0009f48c0, 0x0, 0x0, 0x1b2f601, ...)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/expression_rewriter.go:142 +0x163\ngithub.com/pingcap/tidb/planner/core.(*PlanBuilder).rewrite(0xc001356a80, 0x3f8a0a0, 0xc001400840, 0x3fc1e80, 0xc0013c74d0, 0x3fdd980, 0xc0009f48c0, 0x0, 0x1, 0x0, ...)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/expression_rewriter.go:110 +0xb9\ngithub.com/pingcap/tidb/planner/core.(*PlanBuilder).resolveGbyExprs(0xc001356a80, 0x3f8a0a0, 0xc001400840, 0x3fdd980, 0xc0009f48c0, 0xc0014005d0, 0xc00145c068, 0x1, 0x1, 0x0, ...)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/logical_plan_builder.go:3012 +0x347\ngithub.com/pingcap/tidb/planner/core.(*PlanBuilder).buildSelect(0xc001356a80, 0x3f8a0a0, 0xc001400840, 0xc001309900, 0x3fdd980, 0xc0009f48c0, 0x0, 0x0)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/logical_plan_builder.go:3325 +0x1cce\ngithub.com/pingcap/tidb/planner/core.(*PlanBuilder).buildResultSetNode(0xc001356a80, 0x3f8a0a0, 0xc001400840, 0x3fa8ce0, 0xc001309900, 0x0, 0x0, 0xc001303750, 0x134eb1e)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/logical_plan_builder.go:376 +0x974\ngithub.com/pingcap/tidb/planner/core.(*expressionRewriter).buildSubquery(0xc0013c52b0, 0x3f8a0a0, 0xc001400840, 0xc0013c7560, 0x0, 0x0, 0x0, 0x0)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/expression_rewriter.go:325 +0x9e\ngithub.com/pingcap/tidb/planner/core.(*expressionRewriter).handleScalarSubquery(0xc0013c52b0, 0x3f8a0a0, 0xc001400840, 0xc0013c7560, 0x38, 0xc0002b1dd0, 0x203000)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/expression_rewriter.go:947 +0x6a\ngithub.com/pingcap/tidb/planner/core.(*expressionRewriter).Enter(0xc0013c52b0, 0x3fa9280, 0xc0013c7560, 0xc000316d20, 0xc0013211e8, 0x2)\n\t/home/jenkins/agent/workspace/tidb_master/go/src/github.com/pingcap/tidb/planner/core/expression_rewriter.go:392 +0x91a\ngithub.com/pingcap/parser/ast.(*SubqueryExpr).Accept(0xc0013c7560, 0x3f52260, 0xc0013c52b0, 0x3fa79c0, 0xc0009b4fc0, 0xc0009b5000)\n\t/home/jenkins/agent/workspace/tidb_master/go/pkg/mod/github.com/pingcap/parser@v0."]
```

### 4. What is your TiDB version? (Required)
```yaml
Release Version: v4.0.0-beta.2-2148-g381ac81f2-dirty
Edition: Community
Git Commit Hash: 381ac81f2df09093096c0bda4ba2c9118b759488
Git Branch: master
UTC Build Time: 2021-02-19 14:00:26
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #22850 

Problem Summary: planner: remove some risky cache operations in the plan builder

### What is changed and how it works?

https://github.com/pingcap/tidb/pull/21431/ allows the plan builder to cache some logical plans to avoid unnecessary evaluation(in `PlanBuilder.buildTableRefs`), but this operation brings some risks.
Because the cached logical plans may be modified in some other places, and once they are modified, some implicit assumptions may be broken.
image

In this case(#22850): 
1) the expression rewriter has an implicit assumption that `len(p.Schema().Cols) == len(p.OutputNames())` (in `getExpressionRewriter`). 
2) the subquery in the SQL is handled twice during processing(in `buildProjection`): the first happens when pre-processing user variables(`preprocessUserVarTypes`) and the second happens when rewriting selection fields(`rewriteWithPreprocess`);
3) during the second processing, the result cached in first is returned, but the cached logical plan has been modified since its output names have been pruned, so the implicit assumption required by the rewriter is broken.

### Check List 

Tests 

- Unit test

### Release note 

- planner: remove some risky cache operations in the plan builder

### CODE REVIEW

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -312,20 +312,7 @@ func (b *PlanBuilder) buildTableRefs(ctx context.Context, from *ast.TableRefsCla
...
-		return
-	}
-	p, err = b.buildResultSetNode(ctx, from.TableRefs)
-	if err != nil {
-		return nil, err
-	}
-	b.cachedResultSetNodes[from.TableRefs] = p
-	b.cachedHandleHelperMap[from.TableRefs] = b.handleHelper.tailMap()
-	return
+	return b.buildResultSetNode(ctx, from.TableRefs)
```
Comment (rebelice) [inline planner/core/logical_plan_builder.go:315]: The operation of this line is the same as the above branch without cache, can we merge them, or add a comment

Comment (qw4990) [reply planner/core/logical_plan_builder.go:308]: OK, fixed.

**Review by rebelice — APPROVED** [review]

**Review by winoros — COMMENTED** [review]
lgtm

**Review by winoros — APPROVED** [review]

**Pull request merged**